File: Blinky/nano_blink_manual.c
#include <avr/io.h>




const uint8_t led_pin = 5;
const uint8_t led_bit = 1 << led_pin;

#define loop_ms 0x0a52

void delay(uint16_t ms) {
    uint16_t count = 0;
    while (count < ms) {
        uint16_t loop = 0;
        while (loop < loop_ms) {
            __asm__ __volatile__ ("nop\n\t");
            loop += 1;
        }
        count += 1;
    }
}


int main (void) {
    DDRB |= _BV(DDB5);
    PORTB = led_bit;
    
    
    
    while(1) {
        PORTB ^= led_bit;
        delay(500);
    }
}

File: Blinky/nano_blink_manual.zig
const avr = struct {
    pub const ddrb = @as(*volatile u8, @ptrFromInt(0x24));
    pub const portb = @as(*volatile u8, @ptrFromInt(0x25));
};

const led_pin: u8 = 5;
const led_bit: u8 = 1 << led_pin;

const loop_ms = 0x0a52;

fn delay(ms: u16) void {
    var count: u16 = 0;
    while (count < ms) : (count += 1) {
        var loop: u16 = 0;
        while (loop < loop_ms) : (loop += 1) {
            asm volatile ("" ::: "memory");
            
        }
        
    }
}


export fn main() noreturn {
    avr.ddrb.* = led_bit;
    avr.portb.* = led_bit;
    
    
    
    while (true) {
        avr.portb.* ^= led_bit;
        delay(500);
    }
}

